### DISCUSSION TITLE
请教 skynet的 codecache 问题

### DISCUSSION THREAD

**Discussion by aceyin** [cloudwu/skynet#1095] [Q&A]
## 背景：
    现在 有2个service和一个 util：service_a.lua, service_b.lua, util.lua 
    在 service_a 和 service_b 里面都 require 了 util:
    local util = require 'util'
---
## 问题：
    按照我对 codecache 文档所描述 的理解， 无论是 service_a, service_b 还是 util 在整个 
    skynet 进程里面都只会有一份代码被加载（我这个理解是不是对的？）
    但是在 service_a 和 service_b 里面打印出来的 util 的内存地址是不同的， 看起来就像 
    codecache 没生效一样
---
## 求助：
    1. 是不是我对 codecache 的理解有问题？如果是，codecache 真正的意图是啥？
    2. 如果我的理解没问题， 那skynet每启动1个service， 就会load所有的lua文件一次， 从而 
       并没有达到网上说的 “无论有多少个service（虚拟机）实例，但是lua proto只会有1份“的说法

    请各位不吝赐教

**Comment by hanxi** [comment]
codecache 节省的是代码占用内存，不是运行中使用的内存。

**Comment by aceyin** [comment]
> codecache 节省的是代码占用内存，不是运行中使用的内存。

原来是这样的。。。
那等于说不同 service 里面所require的对象， 都是该service里面新创建的，都需要独自占用一份服务器的内存。。。

感谢回复~

**Comment by findix** [comment]
每个 Service 都是独立的 Lua 虚拟机，如果考虑内存占用问题，大的数据可以使用 sharetable 共享。
